### DISCUSSION TITLE
IRHVAC for KELON168 protocol

### DISCUSSION THREAD

**Discussion by dgel27** [arendst/Tasmota#22006] [Q&A] [answered]
Hi,
I have 3 AC's that used KELON168 protocol. I can receive and send codes of KELON168 with IrSend command:
{"IrReceived":{"Protocol":"KELON168","Bits":168,"Data":"0x83060072000088180000000000E200020000000002","Repeat":0}}
But IRHVAC isn't decode it, and it doen't have it in supported vendor list (supported KELON, but not KELON168).

I have look at IRremoteESP8266 library, and it seems that this protocol present, but in beta version.
Maybe it possible to add some option to compile FW with KELON168 supported?

I will be appreciated for help how to do it.

Thanks,
DG

**Comment by sfromis** [comment]
On an IR bridge flashed with current full IR binary (`tasmota-ir.bin.gz`, which you already need for HVAC), `KELON168` is included in the list of protocols produced when I try to use an unknown protocol.

**Comment by dgel27** [comment]
Hi @sfromis,
Thanks for quich reply,
I just burn original version from web installer [Project tasmota - Tasmota Version 14.2.0(release-ir)-2_7_7(2024-08-14T12:36:29)], and there no KELON168 in the list of IRHVAC vendors, only KELON:

15:52:50.285 CMD: IRHVAC {"Vendor":"AIRLL","Model":-1,"Command":"Control","Mode":"Cool","Power":"On","Celsius":"On","Temp":24,"FanSpeed":"Max","SwingV":"Off","SwingH":"Off","Quiet":"Off","Turbo":"Off","Econo":"Off","Light":"Off","Filter":"Off","Clean":"Off","Beep":"Off","Sleep":-1,"iFeel":"Off","SensorTemp":null}
15:52:50.325 RSL: RESULT = {"IRHVAC":"Wrong Vendor (LG|COOLIX|DAIKIN|KELVINATOR|MITSUBISHI_AC|GREE|ARGO|TROTEC|TOSHIBA_AC|FUJITSU_AC|MIDEA|HAIER_AC|HITACHI_AC|HITACHI_AC1|HAIER_AC_YRW02|WHIRLPOOL_AC|SAMSUNG_AC|ELECTRA_AC|PANASONIC_AC|LG2|DAIKIN2|VESTEL_AC|TECO|TCL112AC|MITSUBISHI_HEAVY_88|MITSUBISHI_HEAVY_152|DAIKIN216|SHARP_AC|GOODWEATHER|DAIKIN160|NEOCLIMA|DAIKIN176|DAIKIN128|AMCOR|DAIKIN152|MITSUBISHI136|MITSUBISHI112|HITACHI_AC424|DAIKIN64|AIRWELL|DELONGHI_AC|CARRIER_AC64|HITACHI_AC344|CORONA_AC|SANYO_AC|VOLTAS|TRANSCOLD|TECHNIBEL_AC|MIRAGE|PANASONIC_AC32|ECOCLIM|TRUMA|HAIER_AC176|TEKNOPOINT|KELON|TROTEC_3550|SANYO_AC88|RHOSS|AIRTON|HITACHI_AC264|HITACHI_AC296|HAIER_AC160|BOSCH144|YORK)"}

Are you using ESP32 device?

I see it present in Tasmota/lib/lib_basic/IRremoteESP8266/IRremoteESP8266/src/IRremoteESP8266.h , but i don't know how to get it work

Thanks,
DG

**Reply by s-hadinger** [reply]
ESP32 and ESP8266 have the exact same list of supported devices.

KELON168 is only supported for raw send/receive like indicated by the frame you received. Example:
```
 Irsend {"Protocol":"KELON168","Bits":168,"Data":"0x83060072000088180000000000E200020000000002"}
```
works fine.

Encoding and decoding of the frame is not supported by IRRemoteESP8266.

**Reply by dgel27** [reply]
Hi @s-hadinger,

Yes, IrSend works fine for me.
I just see in IRRemoteESP8266 library code for encoding and decoding KELON168, and thouth maybe it possibly in some way compile it.
I see #ifdefs for decoding and sending of KELON168 in Tasmota/lib/lib_basic/IRremoteESP8266/IRremoteESP8266/src/ir_Kelon.cpp:

`#if SEND_KELON168
/// Send a Kelon 168 bit / 21 byte message.
/// Status: BETA / Probably works.
/// @param[in] data The data to be transmitted.
/// @param[in] nbytes Nr. of bytes of data to be sent.
/// @param[in] repeat The number of times the command is to be repeated.
void IRsend::sendKelon168(const uint8_t data[], const uint16_t nbytes,
                          const uint16_t repeat) {
  assert(kKelon168StateLength == kKelon168Section1Size + kKelon168Section2Size +
                                 kKelon168Section3Size);
  // Enough bytes to send a proper message?
  if (nbytes < kKelon168StateLength) return;

  for (uint16_t r = 0; r <= repeat; r++) {
    // Section #1 (48 bits)
    sendGeneric(kKelonHdrMark, kKelonHdrSpace,
                kKelonBitMark, kKelonOneSpace,
                kKelonBitMark, kKelonZeroSpace,
                kKelonBitMark, kKelon168FooterSpace,
                data, kKelon168Section1Size, kKelonFreq, false,  // LSB First.
                0,  // No repeats here
                kDutyDefault);
    // Section #2 (64 bits)
    sendGeneric(0, 0,
                kKelonBitMark, kKelonOneSpace,
                kKelonBitMark, kKelonZeroSpace,
                kKelonBitMark, kKelon168FooterSpace,
                data + kKelon168Section1Size, kKelon168Section2Size,
                kKelonFreq, false,  // LSB First.
                0,  // No repeats here
                kDutyDefault);
    // Section #3 (56 bits)
    sendGeneric(0, 0,
                kKelonBitMark, kKelonOneSpace,
                kKelonBitMark, kKelonZeroSpace,
                kKelonBitMark, kKelonGap,
                data + kKelon168Section1Size + kKelon168Section2Size,
                nbytes - (kKelon168Section1Size + kKelon168Section2Size),
                kKelonFreq, false,  // LSB First.
                0,  // No repeats here
                kDutyDefault);
  }
}
#endif  // SEND_KELON168

#if DECODE_KELON168
/// Decode the supplied Kelon 168 bit / 21 byte message.
/// Status: BETA / Probably Working.
/// @param[in,out] results Ptr to the data to decode & where to store the result
/// @param[in] offset The starting index to use when attempting to decode the
///   raw data. Typically/Defaults to kStartOffset.
/// @param[in] nbits The number of data bits to expect.
/// @param[in] strict Flag indicating if we should perform strict matching.
/// @return True if it can decode it, false if it can't.
bool IRrecv::decodeKelon168(decode_results *results, uint16_t offset,
                            const uint16_t nbits, const bool strict) {
  if (strict && nbits != kKelon168Bits) return false;
  if (results->rawlen <= 2 * nbits + kHeader + kFooter * 2 - 1 + offset)
    return false;  // Can't possibly be a valid Kelon 168 bit message.

  uint16_t used = 0;

  used = matchGeneric(results->rawbuf + offset, results->state,
                      results->rawlen - offset, kKelon168Section1Size * 8,
                      kKelonHdrMark, kKelonHdrSpace,
                      kKelonBitMark, kKelonOneSpace,
                      kKelonBitMark, kKelonZeroSpace,
                      kKelonBitMark, kKelon168FooterSpace,
                      false, _tolerance, 0, false);
  if (!used) return false;  // Failed to match.
  offset += used;

  used = matchGeneric(results->rawbuf + offset,
                      results->state + kKelon168Section1Size,
                      results->rawlen - offset, kKelon168Section2Size * 8,
                      0, 0,
                      kKelonBitMark, kKelonOneSpace,
                      kKelonBitMark, kKelonZeroSpace,
                      kKelonBitMark, kKelon168FooterSpace,
                      false, _tolerance, 0, false);
  if (!used) return false;  // Failed to match.
  offset += used;

  used = matchGeneric(results->rawbuf + offset,
                      results->state + (kKelon168Section1Size +
                                        kKelon168Section2Size),
                      results->rawlen - offset,
                      nbits - (kKelon168Section1Size +
                               kKelon168Section2Size) * 8,
                      0, 0,
                      kKelonBitMark, kKelonOneSpace,
                      kKelonBitMark, kKelonZeroSpace,
                      kKelonBitMark, kKelonGap,
                      true, _tolerance, 0, false);
  if (!used) return false;  // Failed to match.

  results->decode_type = decode_type_t::KELON168;
  results->bits = nbits;
  return true;
}
#endif  // DECODE_KELON168
`
Maybe it for IrSend/IrReceiv, and not for IRHVAC.

Thanks,
DG

**Reply by sfromis** [reply]
Well, the key part was that you need to be using a binary with full IR, instead of the basic IR included in most builds. A standard build on ESP32 does not have full IR. The command suggested by s-hadinger works when you are using a build with `#define USE_IR_REMOTE_FULL`, both on ESP8266 and ESP32, but not otherwise.
`Irsend {"Protocol":"KELON168","Bits":168,"Data":"0x83060072000088180000000000E200020000000002"}`

If your ESP32 variant is the basic (oldest) one, the project supplies the build `tasmota32-ir.bin`, which should include full IR.
https://ota.tasmota.com/tasmota32/release/

**Reply by s-hadinger** [reply, marked as answer]
The code section above is for `IrSend` only. What they mean with encode/decode is just about the raw bitstream. There is no class corresponding to actually extracting the meaning of those bytes.

**Reply by dgel27** [reply]
Thanks @sfromis, @s-hadinger.
I see there is no code for AC encodind/decoding.

edit:
I will answer to myself :):
1. i found test branch of IRremoteESP8266 with KELON167 AC support: [https://github.com/leonardfactory/IRremoteESP8266/tree/dev-kelon168-ac-support](https://github.com/leonardfactory/IRremoteESP8266/tree/dev-kelon168-ac-support)
2. I think need to add support also to IRac.h and IRac.cpp files. I don't test it, because:
3. **WHIRLPOOL_AC**  works on KELON167 AC!!!
